File: Week 4/solutions/C_horrorlist.py
from queue import Queue

INF = 10 ** 9 # useful way to keep track of infinity as a really large number

#input format: for each index i, related[i] is a list of indices adjacent to i
def horrorlist(n, horror_movies, related):
    horror_index = [INF] * n
    q = Queue()
    for i in range(len(horror_movies)): 
        horror_index[horror_movies[i]] = 0
        q.put(horror_movies[i])

    while not q.empty():
        curr = q.get()
        for i in related[curr]:
            if horror_index[i] == INF:
                horror_index[i] = horror_index[curr] + 1
                q.put(i)

    highest_score, best_index = -1, 0
    for i in range(n):
        if horror_index[i] > highest_score:
            highest_score = horror_index[i]
            best_index = i

    
    return best_index


n,h,l = map(int, input().split())
horror_movies = list(map(int, input().split()))
related = [ [] for i in range(n) ]
for i in range(l):
    u,v = map(int, input().split())
    related[u].append(v)
    related[v].append(u)

print(horrorlist(n, horror_movies, related))

File: Week 4/solutions/C_horrorlist.cpp

#include <bits/stdc++.h>
using namespace std;

//input format: related has a list of indices adjacent to each given index
//return the integer that has the highest horror index
int horrorlist(int n, vector<int>& horror_movies, vector<vector<int>>& related){
    vector<int> horror_index(n, INT_MAX);
    queue<int> q;
    for(int i : horror_movies){
        horror_index[i] = 0;
        q.push(i);
    }   
    while(q.size()){
        int u = q.front();
        q.pop();
        for(int v : related[u]){
            if(horror_index[v] == INT_MAX){
                horror_index[v] = horror_index[u] + 1;
                q.push(v);
            }
        }
    }
    int best = -1, best_index = -1;
    for(int i = 0; i < n; i++){
        if(horror_index[i] > best){
            best = horror_index[i];
            best_index = i;
        }
    }
    return best_index;
}
int main(){
    int n, h, l;
    cin >> n >> h >> l;
    vector<int> horror_movies(h);
    for(int i = 0; i < h; i++){
        cin >> horror_movies[i];
    }
    vector<vector<int>> related(n);
    for(int i = 0; i < l; i++){
        int u, v;
        cin >> u >> v;
        related[u].push_back(v);
        related[v].push_back(u);
    }
    cout << horrorlist(n, horror_movies, related) << endl;
}